### ISSUE TITLE
Parametrising `contents` byte sequence

### ISSUE THREAD

**Issue by agatti** [kaitai-io/kaitai_struct#984]
Hi there,

I have a file made up of TLV blocks which may or may not be nested.  Since there are a few valid possibilities about blocks' order and nesting, I'm trying to add validation right inside the Kaitai structure file by parametrising the generic TLV block definition and passing the expected type code to the type and using that value as the `contents` value for the type field.

Now, `contents` takes the value as a string, and it looks for the string representation of the enum name, which is not what I need, and maybe I am doing something wrong there.

Alternatively, I've tried to add a `process` directive to have the type check happening in my code, but still driven by Kaitai - so the definition would look like this:

```yaml
# ...

types:
  header_block:
    seq:
      - id: field1
        type: block(types::...)
      - id: field2
        type: block(types::...)

# ...

  block:
    params:
      - id: expected_type
        type: u4
        enum: types
    seq:
      - id: type
        type: u4
        enum: types
        process: field_code_validator(expected_type, type)
      - id: payload_length
        type: u4
      - id: payload
        type:
          switch-on: type
          cases:

# ...
```

But the `process` directive is not being accepted, no matter whether `ks-opaque-types: true` is present in the `meta` section or not:

```
/types/block/seq/0/process:
error: unknown key found, expected: consume, doc, doc-ref, eos-error, id, if, include, repeat, terminator, type, valid
```

Is it possible to achieve this?

**Comment by generalmimon** [comment]
@agatti See https://github.com/kaitai-io/kaitai_struct/issues/435 - you're looking for `valid`:

```diff
   block:
     params:
       - id: expected_type
         type: u4
         enum: types
     seq:
       - id: type
         type: u4
         enum: types
-        process: field_code_validator(expected_type, type)
+        valid: expected_type
       - id: payload_length
         type: u4
```

`valid` still has no proper documentation at https://doc.kaitai.io/user_guide.html, see https://github.com/kaitai-io/kaitai_struct/issues/944, which is probably our biggest documentation debt at the moment and we should fix it ASAP (`valid` was already in 0.9, we have 0.10 now and it's very embarrassing that this issue still persists). Unfortunately, documentation work tends to get less priority than it should ☹️

**Comment by generalmimon** [comment]
> But the `process` directive is not being accepted, no matter whether `ks-opaque-types: true` is present in the `meta` section or not

`process` is meant for "bytes in, bytes out" processing, see https://doc.kaitai.io/user_guide.html#process. You can't use it on a primitive types - only byte arrays and user types [with substream](https://doc.kaitai.io/user_guide.html#_applying_process_without_a_size) are supported.

Primitive types (`type: {uX,sX,bX,fX}`, `enum` only maps the resulting value but doesn't change primitive type behavior) cannot have their own substream at the moment - applying `size` on them does nothing (AFAIR, `size` is still silently allowed and ignored even in the latest 0.10 compiler, but we will no longer allow that in a future KSC version).

So there is no byte array / substream that `process` can be applied to, so the compiler treats `process` as an unknown key (which I agree is not really helpful, because it's only an unknown *in the context* of other keys used - `type: u4` plays a key role in it - but the compiler doesn't give any information whatsoever about this context).

**Comment by Mingun** [comment]
> but we will no longer allow that in a future KSC version

I hope that this will be solved in other way, which is described in #788. `size` is a field trait, not a type trait

**Comment by agatti** [comment]
@generalmimon Thanks a lot, `valid` did work indeed.  My rationale for using `process` was to write a pass-through validator that would assert on a types mismatch.

Regarding documentation, it would be nice if the JSON schema would get brought to date, whenever there's time for that.  I'm also working on a Kaitai plugin for JetBrains' IDEs and I've had to patch things up a bit on that front.  It's a bit embarrassing to note that `valid` was one of the tags I came across when testing the plugin and I totally blanked out on that!

Anyway, thanks again, I guess this can be closed if everybody's ok with that.
